File: Backend/src/main/java/com/example/model/UserProgress.java
package com.example.model;


import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "user_progress")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class UserProgress {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne
    @JoinColumn(name = "user_id", nullable = false, unique = true)
    private User user;

    @Column(nullable = false)
    private Integer totalXp = 0;

    @Column(nullable = false)
    private Integer currentLevel = 1;

    @Column(nullable = false)
    private Integer quizCompleted = 0;

    @Column(nullable = false)
    private Integer quizSucceeded = 0;

    @Column(nullable = false)
    private Integer totalStudyTimeMinutes = 0;

    @Column(nullable = false)
    private Integer videosWatched = 0;

    @Column(nullable = false)
    private Double averageSuccessRate = 0.0;

    @Column(nullable = false)
    private Integer currentStreak = 0;

    @Column(nullable = false)
    private Integer longestStreak = 0;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @Column
    private LocalDateTime lastActivityDate;

    // Méthodes utilitaires
    public void addXp(int xp) {
        this.totalXp += xp;
        updateLevel();
    }

    public void addStudyTime(int minutes) {
        this.totalStudyTimeMinutes += minutes;
    }

    public void incrementQuizCompleted() {
        this.quizCompleted++;
    }

    public void incrementQuizSucceeded() {
        this.quizSucceeded++;
    }

    public void incrementVideosWatched() {
        this.videosWatched++;
    }

    public void updateSuccessRate() {
        if (quizCompleted > 0) {
            this.averageSuccessRate = (quizSucceeded * 100.0) / quizCompleted;
        }
    }

    private void updateLevel() {
        // Formule: niveau = floor(XP / 1000) + 1
        // Chaque niveau nécessite 1000 XP
        int newLevel = (totalXp / 1000) + 1;
        if (newLevel > currentLevel) {
            this.currentLevel = newLevel;
        }
    }

    public int getXpForNextLevel() {
        return currentLevel * 1000;
    }

    public int getXpProgressInCurrentLevel() {
        return totalXp % 1000;
    }

    public double getProgressPercentage() {
        return (getXpProgressInCurrentLevel() * 100.0) / 1000;
    }
}

File: Frontend/lib/models/user_progress.dart
class UserProgress {
  final int userId;
  final String userName;
  final String userEmail;
  final String niveau;
  final int totalXp;
  final int currentLevel;
  final int quizCompleted;
  final int quizSucceeded;
  final int totalStudyTimeMinutes;
  final int videosWatched;
  final double averageSuccessRate;
  final int xpForNextLevel;
  final int xpProgressInCurrentLevel;
  final double progressPercentage;
  final int currentStreak;
  final int longestStreak;
  final DateTime? lastActivityDate;
  final DateTime createdAt;
  final String studyTimeFormatted;
  final String levelTitle;

  UserProgress({
    required this.userId,
    required this.userName,
    required this.userEmail,
    required this.niveau,
    required this.totalXp,
    required this.currentLevel,
    required this.quizCompleted,
    required this.quizSucceeded,
    required this.totalStudyTimeMinutes,
    required this.videosWatched,
    required this.averageSuccessRate,
    required this.xpForNextLevel,
    required this.xpProgressInCurrentLevel,
    required this.progressPercentage,
    required this.currentStreak,
    required this.longestStreak,
    this.lastActivityDate,
    required this.createdAt,
    required this.studyTimeFormatted,
    required this.levelTitle,
  });

  factory UserProgress.fromJson(Map<String, dynamic> json) {
    return UserProgress(
      userId: json['userId'] ?? 0,
      userName: json['userName'] ?? '',
      userEmail: json['userEmail'] ?? '',
      niveau: json['niveau'] ?? '',
      totalXp: json['totalXp'] ?? 0,
      currentLevel: json['currentLevel'] ?? 1,
      quizCompleted: json['quizCompleted'] ?? 0,
      quizSucceeded: json['quizSucceeded'] ?? 0,
      totalStudyTimeMinutes: json['totalStudyTimeMinutes'] ?? 0,
      videosWatched: json['videosWatched'] ?? 0,
      averageSuccessRate: (json['averageSuccessRate'] ?? 0).toDouble(),
      xpForNextLevel: json['xpForNextLevel'] ?? 1000,
      xpProgressInCurrentLevel: json['xpProgressInCurrentLevel'] ?? 0,
      progressPercentage: (json['progressPercentage'] ?? 0).toDouble(),
      currentStreak: json['currentStreak'] ?? 0,
      longestStreak: json['longestStreak'] ?? 0,
      lastActivityDate: json['lastActivityDate'] != null
          ? DateTime.parse(json['lastActivityDate'])
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      studyTimeFormatted: json['studyTimeFormatted'] ?? '0min',
      levelTitle: json['levelTitle'] ?? 'Novice',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'userName': userName,
      'userEmail': userEmail,
      'niveau': niveau,
      'totalXp': totalXp,
      'currentLevel': currentLevel,
      'quizCompleted': quizCompleted,
      'quizSucceeded': quizSucceeded,
      'totalStudyTimeMinutes': totalStudyTimeMinutes,
      'videosWatched': videosWatched,
      'averageSuccessRate': averageSuccessRate,
      'xpForNextLevel': xpForNextLevel,
      'xpProgressInCurrentLevel': xpProgressInCurrentLevel,
      'progressPercentage': progressPercentage,
      'currentStreak': currentStreak,
      'longestStreak': longestStreak,
      'lastActivityDate': lastActivityDate?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'studyTimeFormatted': studyTimeFormatted,
      'levelTitle': levelTitle,
    };
  }
}